### PR TITLE
Fixed FP to FP8 downcast

### LINKED ISSUE #4630: Several `python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping` test cases failed
From https://github.com/intel/intel-xpu-backend-for-triton/pull/4621:

From pytest:
```bash
FAILED python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping[float32-float8e4nv-nan] - AssertionError: assert tensor(False, device='xpu:0')
FAILED python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping[float32-float8e5-inf] - AssertionError: Tensor-likes are not 'equal'!
FAILED python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping[float32-float8e5--inf] - AssertionError: Tensor-likes are not 'equal'!
FAILED python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping[float16-float8e4nv-nan] - AssertionError: assert tensor(False, device='xpu:0')
FAILED python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping[float16-float8e5-inf] - AssertionError: Tensor-likes are not 'equal'!
FAILED python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping[float16-float8e5--inf] - AssertionError: Tensor-likes are not 'equal'!
FAILED python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping[bfloat16-float8e4nv-max] - AssertionError: Tensor-likes are not 'equal'!
FAILED python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping[bfloat16-float8e4nv-min] - AssertionError: Tensor-likes are not 'equal'!
FAILED python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping[bfloat16-float8e4nv-inf] - AssertionError: Tensor-likes are not 'equal'!
FAILED python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping[bfloat16-float8e4nv--inf] - AssertionError: Tensor-likes are not 'equal'!
FAILED python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping[bfloat16-float8e4nv-nan] - AssertionError: assert tensor(False, device='xpu:0')
FAILED python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping[bfloat16-float8e5-inf] - AssertionError: Tensor-likes are not 'equal'!
FAILED python/test/unit/language/test_conversions.py::test_typeconvert_downcast_clamping[bfloat16-float8e5--inf] - AssertionError: Tensor-likes are not 'equal'!
```

Maybe useful as reference: https://github.com/triton-lang/triton/commit/20a8ac9945c4dcdce2991331b9f65377b15a588f#diff-c1b7645c56652fe811ae807f37bbea185af4221b1122fa095fbd43b80266182e

### PR DESCRIPTION
**Pull request by AndreyPavlenko**
Fixes #4630

This implementation is based on an arithmetic approach (see below) and provides an RTNE conversion from F32/16 to F8 types. Despite it is using more expensive fp binary operations, I don't see any performance degradation (you can find a simple bench commented out in the test). The implementation is very simple and template-based. It replaces 4 existing functions that are more complicated and adds a direct conversion from F32 to F8 without intermediate F16.

The implementation is based on the following approach. To convert a float from src to dst, we need to find such an exponent and mantissa that src is equal or close to dst. For the normal numbers (exp != 0) it means:
```python
2^(srcExp - SrcBias) * (1 + srcMan/2^SrcMBits) = 2^(dstExp - DstBias) * (1 + dstMan/2^DstMBits)
```
and the following for subnormals (dstExp == 0):
```python
src = 2^(1 - DstBias) * (dstMan/2^DstMBits)
```
The exponent is calculated as:
```python
dstExp = max(0, srcExp - SrcBias + DstBias)
```
Simplifying the first formula, we can get the following:
```python
dstMan = srcMan * 2^(DstMBits - SrcMBits)
```
If `SrcBias == DstBias` (in the case of FP16 to F8E5 conversion), this formula also works for subnormals. In the general case, we can use the following formula for subnormals:
```python
dstMan = src * 2^(DstMBits + DstBias - 1)
```
Thus, to get the mantissa, we are using a simple multiplication by a constant and rounding the result to the nearest int. In case of the mantissa overflow, we need to reset it to zero and increment the exponent. It could be done in the following way:
```python
dst = max((dstExp << DstMBits) + dstMan, DST_MAX)
```
It gives us the required dst value without the sign bit.

### CODE REVIEW

File: python/test/unit/intel/test_conversions.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+@triton.jit
+def type_convert(src, dst, rounding: tl.constexpr, BLOCK_SIZE: tl.constexpr):
+    idxs = tl.program_id(0) * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    x = tl.load(src + idxs)
+    y = x.to(dst.dtype.element_ty, fp_downcast_rounding=rounding)
+    tl.store(dst + idxs, y)
+
+
+@pytest.mark.parametrize("dst_dtype", ["float8e4nv", "float8e5"])
```
Comment (kurapov-peter) [inline python/test/unit/intel/test_conversions.py]: nit: You can parameterize with actual types and drop the string to type conversion.

File: python/test/unit/intel/test_conversions.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+@pytest.mark.parametrize("src_dtype", ["float16", "bfloat16", "float32"])
+def test_convert_to_fp8(src_dtype, dst_dtype, device):
+    src_dtype = NAME_TO_DTYPE[src_dtype]
+    dst_dtype = NAME_TO_DTYPE[dst_dtype]
+    src_idtype = torch.int32 if src_dtype == torch.float32 else torch.int16
+    max_val = torch.finfo(dst_dtype).max
+    min_val = torch.tensor(-1.0 * max_val, dtype=dst_dtype).view(torch.uint8).item()
```
Comment (kurapov-peter) [inline python/test/unit/intel/test_conversions.py]: This looks a bit suspicious, does `-1 * max_val` really give the minimum value? `torch.finfo(dst_dtype).min` would be safer

File: python/test/unit/intel/test_conversions.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    SIZE = 2**16
+    BLOCK_SIZE = SIZE // 32
+    src = torch.arange(0, SIZE, dtype=src_idtype, device=device)
+    if src_dtype == torch.float32:
+        src = src << 16 | src
+    src = src.view(src_dtype)
+    dst = torch.empty_like(src, dtype=dst_dtype, device=device)
+
+    def run_kernel():
+        type_convert[(SIZE // BLOCK_SIZE, )](triton.reinterpret(src, src_dtype), triton.reinterpret(dst, dst_dtype),
```
Comment (kurapov-peter) [inline python/test/unit/intel/test_conversions.py]: just curious, why do you need to reinterpret the destination?

File: python/test/unit/language/test_conversions.py [hunk 4]
```diff
@@ -234,17 +234,9 @@ def downcast_test(src_dtype, dst_dtype, rounding, exponent_bits, mantissa_bits,
...
-    # back to FP32. This will need to be changed back when HW FP32->FP8 convertion is used for XPU.
-    if device=='xpu' and src_dtype.primitive_bitwidth == 32 and dst_dtype.primitive_bitwidth == 8:
-        src = launch_type_convert_triton(src, src_dtype, tl.float16, device=device, rounding=rounding)
-        src = launch_type_convert_triton(src, tl.float16, tl.float32, device=device)
-    else:
-        src = launch_type_convert_triton(src, src_dtype, tl.float32, device=device)
+    src = launch_type_convert_triton(src, src_dtype, tl.float32, device=device)
 
-    dst2 = launch_downcast_emulated(src, src_dtype, dst_dtype, rounding, exponent_bits, mantissa_bits, exponent_bias, device=device)
+    dst2 = launch_downcast_emulated(src, tl.float32, dst_dtype, rounding, exponent_bits, mantissa_bits, exponent_bias, device=device)
```
Comment (kurapov-peter) [inline python/test/unit/language/test_conversions.py:239]: Could you please explain the type change here?

**Comment by kurapov-peter** [comment]
> I don't see any performance degradation (you can find a simple bench commented out in the test).

I'd expect to see degradation (if any) on a more compute-intensive workload. I think this at least needs to go through the benchmark set that we have.

File: python/test/unit/language/test_conversions.py (hunk 4 above)
Comment (AndreyPavlenko) [reply python/test/unit/language/test_conversions.py:239]: This is how it's implemented upstream https://github.com/triton-lang/triton/blob/e8c4711ec5886957d31067d0ef59b456cac15dc6/python/test/unit/language/test_conversions.py#L215. I've just reverted changes from this https://github.com/intel/intel-xpu-backend-for-triton/pull/464 commit. Now F32->F8 conversion is implemented directly, without intermediate conversion through F16.

File: python/test/unit/intel/test_conversions.py (hunk 3 above)
Comment (AndreyPavlenko) [reply python/test/unit/intel/test_conversions.py]: This test is based on the upstream https://github.com/intel/intel-xpu-backend-for-triton/blob/main/python/test/unit/language/test_conversions.py#L455 . The parameter names and the min/max definitions are copied from there.

Comment (AndreyPavlenko) [reply python/test/unit/intel/test_conversions.py]: Without it, the test fails with `TypeError: Unsupported type: <class 'torch.dtype'>`

**Comment by AndreyPavlenko** [comment]
> I think this at least needs to go through the benchmark set that  have.

Here are the bench results - https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/16470155352/job/46557072547

**Comment by leonling-ll** [comment]
> > I think this at least needs to go through the benchmark set that  have.
> 
> Here are the bench results - https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/16470155352/job/46557072547

It's not obvious to see the performance change from the existing benchmark. Can you add the fp to fp8 cases to [benchmarks/micro_benchmarks/conversion/float_conversion/float_conversion.py](https://github.com/intel/intel-xpu-backend-for-triton/blob/main/benchmarks/micro_benchmarks/conversion/float_conversion/float_conversion.py)? So that we can compare the downcasting performance before and after this change dirrectly.

File: third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp [hunk 5]
```diff
@@ -524,60 +442,69 @@ static SmallVector<Value> Fp16_to_Fp8E4M3Nv(Location loc,
...
+          int64_t SrcEBits = std::is_same_v<SrcTy, Float16Type> ? 5 : 8,
+          int64_t DstEBits = std::is_same_v<DstTy, Float8E4M3Type> ? 4 : 5,
+          int64_t SrcMBits = SrcBits - SrcEBits - 1,
+          int64_t DstMBits = 8 - DstEBits - 1,
+          int64_t SrcBias = (1L << (SrcEBits - 1)) - 1,
+          int64_t DstBias = (1L << (DstEBits - 1)) - 1>
+static SmallVector<Value> Fp_to_Fp8_RTNE(Location loc,
+                                         ConversionPatternRewriter &rewriter,
+                                         const SmallVector<Value> &v) {
+  static_assert((std::is_same_v<SrcTy, Float16Type>) ||
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp:456]: Suggestion: can you use `llvm,::is_one_of` as in:
```
static_assert(llvm::is_one_of<SrcTy, Float16Type, BFloat!6Type, Floaf32Type>::value);

File: third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp [hunk 6]
```diff
@@ -524,60 +442,69 @@ static SmallVector<Value> Fp16_to_Fp8E4M3Nv(Location loc,
...
+  static_assert((std::is_same_v<DstTy, Float8E4M3Type>) ||
+                (std::is_same_v<DstTy, Float8E5M2Type>));
+  constexpr int64_t SRC_MASK = (1L << (SrcBits - 1)) - 1;
+  constexpr int64_t SRC_MMASK = (1L << SrcMBits) - 1;
+  constexpr int64_t DST_NAN = 0x7F;
+  constexpr int64_t DST_MAX =
+      std::is_same_v<DstTy, Float8E5M2Type> ? 0x7B : DST_NAN - 1;
+  constexpr bool IS_FP16 = std::is_same_v<SrcTy, Float16Type>;
+  TritonLLVMIRRewriter b(loc, rewriter);
+  auto srcTy = IS_FP16 ? f16_ty : f32_ty;
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp]: When `IS_FP16` is true `SrcTy` may be either `Float32Type` or `BFloat16Type`. So why is `srcTy` == f32_ty in that case ?

File: third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp [hunk 7]
```diff
@@ -524,60 +442,69 @@ static SmallVector<Value> Fp16_to_Fp8E4M3Nv(Location loc,
...
+  auto srcITy = rewriter.getIntegerType(SrcBits);
+  auto fval = [&b](float v) { return IS_FP16 ? b.f16_val(v) : b.f32_val(v); };
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp:470]: When `IS_FP16` is false why return an fp32 value, I wonder whether we should return a bf16 value when `SrcTy` is `Float16Type`

File: third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp (hunk 6 above)
Comment (AndreyPavlenko) [reply third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp]: It's mentioned below, that bf16 is converted to fp32, because `llvm.nearbyint` does not support bf16.

**Comment by AndreyPavlenko** [comment]
> Can you add the fp to fp8 cases to [benchmarks/micro_benchmarks/conversion/float_conversion/float_conversion.py](https://github.com/intel/intel-xpu-backend-for-triton/blob/main/benchmarks/micro_benchmarks/conversion/float_conversion/float_conversion.py)?

Done. I've also modified the benchmark and replaced the backward conversions to fp32 in the loop with bitcasts and arithmetics. Performance impact from these operations should be comparable for all the benchmarks, unlike the conversion to fp32, which is different for each type and, for example, in case of fp8e4->fp32 the impact is significant.

Here are the results:
Before (https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/16583834920/job/46905289067):
```TSV
            N  fp16->fp8e4nv/rtne  fp16->fp8e5/rtne  bf16->fp8e4nv/rtne  bf16->fp8e5/rtne  fp32->fp8e4nv/rtne  fp32->fp8e4nv/rtz  fp32->fp8e5/rtne  fp32->fp8e5/rtz  fp32->fp16/rtne  fp32->fp16/rtz  fp32->bf16/rtne  fp32->bf16/rtz
0      4096.0            0.024335          0.071409            0.020754          0.021179            0.028802           0.200196          0.052232         0.347119         0.431158        0.409600         0.435745        0.136170
1     16384.0            0.096105          0.282094            0.082199          0.084124            0.113447           0.799220          0.205468         1.388475         1.710230        1.644980         1.721008        0.541441
2     65536.0            0.384060          1.124502            0.328402          0.336220            0.453285           3.187549          0.820431         5.461333         6.812474        6.514513         6.812474        2.161478
3    262144.0            1.533723          4.488767            1.312032          1.344328            1.811888          12.663961          3.279259        21.772757        26.914168       25.852465        27.025155        8.611827
4   1048576.0            5.807355         11.358059            4.205406          4.244560            7.097441          37.130878         12.400378        69.626560        92.630389       87.381333        91.180522       30.183535
5   4194304.0           11.466740         17.374913            6.930443          6.981663           21.496023          61.590367         31.013783       122.928019       155.459748      153.188605       157.918072       52.415696
6  16777216.0           11.705633         17.926291            7.017700          7.070939           22.303010          66.057233         32.718156       142.639143       205.477232      184.243532       208.983757       57.258169
7  67108864.0           11.965820         18.072081            7.039962          7.094290           22.612251          67.263570         33.203800       156.430918       231.905674      206.851598       218.240208       59.060147
```

After (https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/16583855278/job/46905344649):
```TSV
            N  fp16->fp8e4nv/rtne  fp16->fp8e5/rtne  bf16->fp8e4nv/rtne  bf16->fp8e5/rtne  fp32->fp8e4nv/rtne  fp32->fp8e4nv/rtz  fp32->fp8e5/rtne  fp32->fp8e5/rtz  fp32->fp16/rtne  fp32->fp16/rtz  fp32->bf16/rtne  fp32->bf16/rtz
0      4096.0            0.029750          0.040474            0.031517          0.031527            0.046572           0.200784          0.046577         0.347119         0.431158        0.411245         0.432068        0.135809
1     16384.0            0.118725          0.161642            0.124461          0.124461            0.186013           0.801566          0.186055         1.388475         1.713808        1.648290         1.717400        0.542158
2     65536.0            0.474623          0.646057            0.497389          0.497540            0.743629           3.196878          0.743713         5.516498         6.812474        6.553600         6.826667        2.164333
3    262144.0            1.894652          2.576101            1.986541          1.985939            2.967444          12.676209          2.968116        21.772757        26.804090       25.750884        26.694908        8.611827
4   1048576.0            7.011072          9.525581            6.795697          6.793055           10.877344          36.663497         10.877344        68.894612        91.980351       87.381333        91.498778       29.993593
5   4194304.0           13.997811         17.944314           11.900760         11.896710           25.420024          61.464009         25.438525       122.784075       155.922082      153.076788       157.562134       52.402599
6  16777216.0           14.349312         18.512994           12.149303         12.149655           26.614448          66.070240         26.599680       142.590651       206.209636      184.081808       208.697798       57.307064
7  67108864.0           14.455391         18.678501           12.219430         12.220097           26.934693          67.263570         26.937287       154.468556       213.389500      206.908997       217.222969       59.071584
```
There is a slight performance degradation in fp32->fp8e5, but in all other cases the new implementation is faster, especially in case of bf16. Despite the new impl is slower for fp32->fp8e5, it handles properly the corner cases with min/max/inf/nan and the tests #4630 pass. I'm not sure if the old impl will be still faster, when fixed.

File: python/test/unit/intel/test_conversions.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+                                             'rtne', BLOCK_SIZE)
+        torch.xpu.synchronize(device)
+
+    run_kernel()
+
+    # import timeit
+    # time = timeit.timeit(run_kernel, number=100000)
+    # print(f"{src_dtype} -> {dst_dtype} time: {time}.")
```
Comment (leonling-ll) [inline python/test/unit/intel/test_conversions.py]: Remove these.

Comment (AndreyPavlenko) [reply python/test/unit/intel/test_conversions.py]: Done.
